### DISCUSSION TITLE
Forest 🌲 Working Groups

### DISCUSSION THREAD

**Discussion by zettai-caiman** [ChainSafe/forest#2337] [Ideas]
We have a lot of multi dimensional work around Forest 🌲 As the team has gotten bigger, and is expanding @lemmih @LesnyRumcajs and myself are looking at the idea of establishing working groups. The working groups will enable more ownership, autonomy and a sense of confidence in a specific area among our Forest 🌲 staff. 

Currently, these are the categories on where a working group can be established:

- Forest 🌲 Core involving network upgrades, performance & syncing improvements, testing infra, calibration network maintenance, feature development such as retrieval markets, debugging the node, remodelling the architecture. 

- Forest 🌲 Light Client involving research into how to make Forest 🌲 extremely lean, proposing FIPs for Light Client interface, enable FVM developers. 

- Forest 🌲 Tooling involving building use cases around Forest 🌲 such as a block explorer, measure network statistics for Forest 🌲 vs Lotus 𑁍 node usage.  

- Builtin actors task force to support ongoing work around the actors. 

cc: @q9f

**Comment by LesnyRumcajs** [comment]
Thanks for starting the discussion, @lerajk. Given it is getting more challenging to keep up with everything happening directly in the Forest, I believe it's a must, sooner or later, to develop some specialisations and dedicated areas of responsibility. Plus, there's Conway's Law. ;)

![image](https://user-images.githubusercontent.com/9642092/206661596-66970dc6-847a-427f-8294-a65431c4817d.png)

I am not sure about the structure, though. Forest Core seems extremely tightly coupled with Forest Light Client. Forest Tooling may need further clarification. Builtin-Actors is, well, a task force unless we want (or are expected) to work on it full-time and are the partial owners of the repo.

I am not convinced what precisely this structure should be. Following Conway's Law, to my understanding, it may be best first to develop an architecture (with Forest, Light Clients, Infrastructure, Tooling, and Builtin Actors considered in it) and form subteams around it; otherwise, we'll end up with an architecture by chance which isn't always great.

On top of that, to reduce silos (and increase the bus factor), everyone should have at least two areas to look into. For most, it would be Forest Core, but it seems too broad in its current form.

Perhaps we should have a brainstorming session with everyone involved.

**Reply by zettai-caiman** [reply]
Yes @LesnyRumcajs agreed, gonna be one of the top priorities for Forest in Q1 2023. 

Agreed on the points you have made, we definitely need to brainstorm on this first. I'm not sure how to divide the working groups at the moment either. We could possibly start slower with two(2) additional working groups besides Forest Core. Perhaps maintaining the `Forest IaC` and working on the `builtin actors` can be other additional sub-groups. We may not have a lot currently with the builtin actors, but we need to increase our contribution to that area along with FIP engagements.

**Comment by q9f** [comment]
Not unrelated to this discussion: there is always the possibility to create a dedicated workgroup/team focussing on FVM/Actors upstream contributions and maintenance. This would be deeply integrated with the core team but we might have dedicated stream leads for this at some point. Let me know if this is interesting or wrong.

**Reply by LesnyRumcajs** [reply]
I have a feeling (perhaps a wrong one) that this is expected of us. We are supposed to contribute and take a more active part in the Filecoin Core Devs' endeavours, protocol internals and FIPs. At the same time, we may lack the necessary expertise because, well, doing some heavy-duty engineering with Rust asynchronous tasks makes you a better (Rust) developer. Still, it doesn't add to any deep Filecoin-related discussions.

It would be great first to make sure we are all aligned when it comes to stakeholders' expectations.

**Reply by zettai-caiman** [reply]
@q9f Yes, the entire idea of working groups is similar to the streams at ChainSafe. 

Like @LesnyRumcajs mentioned, we need to take a more active engagement with Filecoin protocol development. If we can set up dedicated streams with leads within the Forest core group, we will be in a better position to address the current shortcomings.

**Reply by LesnyRumcajs** [reply]
@lerajk can you elaborate on the `streams`? I'm not sure I stumbled upon this so far (except for IO streams 🤓)

**Reply by q9f** [reply]
We will get a new organizational structure next year that introduces `pillars` (here: `ecosystem`), `programs` (here: `protocol`) and `streams` (here: `forest`, maybe also `fvm` in future). This was supposed to be announced in the last town hall. Not sure, maybe it will be announced this week.

However, this does not touch the existing _operational_ structure of the organization and will likely not have many effects on your day-to-day work on Forest. I didn't even realize I used the term _stream_ here.
